### ISSUE TITLE
class-methods-use-this rule doesn't consider existence of getters and setters

### ISSUE THREAD

**Issue by sznowicki** [eslint/eslint#7911]
Currently class-methods-use-this rule is ignoring existence of getters and setters in javascript.

Let's consider an approach to have proper getter and setter for an object.
```js
// originally bar was a storage engine
let bar;

class Foo {
  get bar() {
    return bar;
  }
  set bar(value) {
    bar = value;
  }
}
```

After that, eslint was starting to complain that get bar() should use this. 

I read the manual which says if method doesn't use this, it should be a static method.
Which doesn't make much sense for getters, since static getter won't be executed for instantiated class. 

Then, to comply to eslint rule, I'll need to not use getter or setter and stick to the setBar(), getBar() methods which can be fine, but I think eslint should be wise enough to detect getters special type of functions. 

I checked the rule code, and it seems like the isInstanceMethod() function should be changed.

**Comment by platinumazure** [comment]
Hmm. On the one hand, I can understand that syntactically, it's not really possible to define a static getter or setter that work on an instance of the class. On the other hand, it feels like this is exactly what the rule is warning about-- methods that don't use `this` should be static anyway. It's a bit weird having a getter that just references something in-scope but not inside the instance.

That said, I don't actually know ES6 all that well, so maybe other team members can step in and decide if this really should be a rule enhancement (or bug?) or not.

**Comment by sznowicki** [comment]
Sure, I'll be happy to get another opinion. For me, the problem is that in current state of the art, the only way to use this in getters or setters is to refer to another property or outside scope.

This, won't work of course:
```
class Foo {
  get bar() {
    return this.bar;
  }
  set bar(v) {
    this.bar = v;
  }
}
```

This will complain about underscore usage:
```
class Foo {
  get bar() {
    return this._bar;
  }
  set bar(v) {
    this._bar = v;
  }
}
```

The only way to comply is to use getters and setters with another property, like this:
```
class Foo {
  constructor() {
    this.x = 1;
    this.y = 2;
  }
  get xy() {
    return this.x + this.y;
  }
  set xy() {
    // can't imagine good use case right now
  }
}
```

**Comment by nzakas** [comment]
I think the rule is working as intended, as this is an edge case. I'd say the only possibility here is add an option that allows you to opt out of checking accessor properties.

**Comment by javlund** [comment]
When using the Page Object pattern with webdriver.io functional tests, one is encouraged to use getters to retrieve elements, even though these are not properties on the class (see http://webdriver.io/guide/testrunner/pageobjects.html).

The majority of our linting errors stem from this, making the rule almost unusable for us, so it would be really nice to have the option of excluding getters.

**Comment by BridgeAR** [comment]
When extending Errors you want to add a getter for the name and maybe also add some attributes that you do not want to be allowed to be changed at all. So you wouldn't add a setter for these entries.

```js
class CustomError extends Error {
  get name() {
    return 'CustomError'; // Could also be written as: this.constructor.name
  }
  get code() {
    return 400; // Static value that should not be changed
  }
}
```

**Comment by sznowicki** [comment]
@BridgeAR 

I think this.constructor.name is not same as class Foo { get name()}. It's this.name in that case, but as you probably guess I agree with you in the matter that it's not a real edge case.

Also, class with getters and setters for same property name will fail in current ES6 version. For the first setter execution, it will jump into an infinite loop.

**Comment by BridgeAR** [comment]
@sznowicki 

```js
// With this.constructor.name
const err = new CustomError('foo')
console.log(err.name) // 'CustomError'

// With this.name
const err = new CustomError('foo')
// <error: RangeError: Maximum call stack size exceeded>
```

**Comment by sznowicki** [comment]
@BridgeAR 

You're extending Error which has setters, that's why you jump into an infinite loop.  Name is also a class name in this example.

This is what I meant: https://gist.github.com/sznowicki/9b3a88efefdb6f899629999e9c9c7457

But, let's not pollute issue with off-topic discussions ;)

**Comment by ljharb** [comment]
I think that in your original example, the fact that it's an ES5 getter/setter is actually irrelevant - the same would have applied in a `getFoo` and `setFoo` method pair.

In other words, these are methods - if methods don't require `this` access, and you want `class-methods-use-this` enabled, then they need override comments.

I could potentially see an option that allowed "class methods that don't use `this`, but assign to a variable that's in scope", but while that would handle your example, it wouldn't handle "a storage engine" as your example's comment indicates.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
